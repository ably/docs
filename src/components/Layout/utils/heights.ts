/*
  Since we perform CSS calc operations involving various UI element heights, we need to centralise
  these values to prevent magic numbers popping up with no obvious reasoning. When making alterations
  to Layout components, consider these values and update where necessary.
*/

export const LANGUAGE_SELECTOR_HEIGHT = 38;
export const INKEEP_ASK_BUTTON_HEIGHT = 96;
