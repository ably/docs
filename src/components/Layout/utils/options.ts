import { LayoutOptions } from 'src/contexts/layout-context';

const defaultLayoutOptions: Record<string, LayoutOptions> = {
  '/docs': { sidebar: false, searchBar: false, template: 'index' },
  '/docs/api/control-api': { sidebar: false, searchBar: true, template: 'control-api' },
  '/docs/sdks': { sidebar: false, searchBar: true, template: 'sdk' },
  '/docs/examples': { sidebar: false, searchBar: true, template: 'examples' },
};

export const getLayoutOptions = (pathname: string) => {
  const cleanPathname = pathname.replace(/\/+$/, '');
  return defaultLayoutOptions[cleanPathname] || { sidebar: true, searchBar: true, template: 'base' };
};
