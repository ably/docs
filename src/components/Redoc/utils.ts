import throttle from 'lodash.throttle';

/*
  This function introduces click listeners to override the functionality for each menu item in the sidebar,
  and also listens for hash changes via scroll to update the currently opened menu section.
  Sadly, this is necessary because Redoc opens the wrong menu item when clicking on a menu trigger, and is a bit
  broken in general.
*/
export const overrideMenuItemNavigation = () => {
  let trackedSection = '';
  const menuItems = document.querySelectorAll('#redoc-container .menu-content .scrollbar-container > ul > li');

  const scrollHandler = throttle(() => {
    const params = new URLSearchParams(window.location.search);
    if (params.get('clicked') === 'true') {
      setTimeout(() => {
        params.delete('clicked');
        window.history.replaceState({}, '', window.location.pathname + '?' + params.toString());
      }, 500);
      return;
    }

    const decodedHash = decodeURIComponent(window.location.hash).split('/')?.[1];

    if (decodedHash !== trackedSection) {
      trackedSection = decodedHash;

      menuItems.forEach((item) => {
        const ul = item.querySelector('ul');
        const svg = item.querySelector('svg');

        if (ul) {
          ul.style.display = '';
        }
        if (svg) {
          svg.style.transform = '';
        }
      });
    }
  }, 500);

  document.addEventListener('scroll', scrollHandler);

  const clickHandlers = new Map();

  menuItems.forEach((targetItem) => {
    const clickHandler = () => {
      // Add clicked query param
      const url = new URL(window.location.href);
      url.searchParams.set('clicked', 'true');
      window.history.replaceState({}, '', url);

      menuItems.forEach((item) => {
        const isOther = item !== targetItem;
        const ul = item.querySelector('ul');
        const svg = item.querySelector('svg');

        if (ul) {
          ul.style.display = isOther ? 'none' : 'block';
        }
        if (svg) {
          svg.style.transform = isOther ? 'rotate(-90deg)' : 'rotate(0deg)';
        }
      });
    };

    clickHandlers.set(targetItem, clickHandler);
    targetItem.addEventListener('click', clickHandler);
  });

  // Clean up function to remove all event listeners
  const cleanup = () => {
    document.removeEventListener('scroll', scrollHandler);
    menuItems.forEach((item) => {
      const handler = clickHandlers.get(item);
      if (handler) {
        item.removeEventListener('click', handler);
      }
    });
    clickHandlers.clear();
  };

  // Watch for URL changes
  const observer = new MutationObserver(() => {
    if (window.location.pathname !== '/docs/api/control-api') {
      cleanup();
    }
  });

  const body = document.querySelector('body');
  if (body) {
    observer.observe(body, { childList: true, subtree: true });
  }

  return cleanup;
};
