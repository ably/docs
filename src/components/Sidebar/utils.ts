import { NavData, NavProductPages, NavProductKey } from './types';

// Determine the active page based on the target link
export const determineActivePage = (data: NavData, targetLink: string): number[] | null => {
  const determinePagePresence = (pages: NavProductPages[], path: number[]): number[] | null => {
    for (let i = 0; i < pages.length; i++) {
      const page = pages[i];

      // If the page is a link and the link matches the target link, return the path
      if ('link' in page && page.link === targetLink) {
        return [...path, i];
      }

      // If the page is a group of pages, recursively check each page
      if ('pages' in page && page.pages) {
        const result = determinePagePresence(page.pages, [...path, i]);
        if (result) {
          return result;
        }
      }
    }
    return null;
  };

  // Iterate through each product and check if the target link is present in the product
  for (const key in data) {
    if (data[key as NavProductKey].content) {
      const result = determinePagePresence(data[key as NavProductKey].content, []);
      if (result) {
        return [Object.keys(data).indexOf(key), ...result];
      }
    }
  }
  return null;
};
