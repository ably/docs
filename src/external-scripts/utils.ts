const defaultAttrs = {
  async: true,
  defer: true,
  data: {},
};

const scriptLoader = (doc, url, customAttrs) => {
  const attrs = { ...defaultAttrs, ...customAttrs };
  const { data } = attrs;
  delete attrs.data;

  const scriptTag = doc.createElement('script');

  Object.keys(attrs).forEach((attr) => (scriptTag[attr] = attrs[attr]));
  scriptTag.src = url;
  scriptTag.type = 'text/javascript';

  Object.keys(data).forEach((attr) => (scriptTag.dataset[attr] = data[attr]));

  // Respect where scripts are first loaded on the page head/end of body etc
  const scripts = doc.getElementsByTagName('script')[0];
  scripts.parentNode.insertBefore(scriptTag, scripts);
};

export { scriptLoader };
