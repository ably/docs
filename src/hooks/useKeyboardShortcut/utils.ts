export const overrideSystemHandling = (e: KeyboardEvent) => {
  if (e) {
    if (e.preventDefault) {
      e.preventDefault();
    }
    if (e.stopPropagation) {
      e.stopPropagation();
    } else if (window.event) {
      window.event.cancelBubble = true;
    }
  }
};

// Function stolen from this Stack Overflow answer:
// https: stackoverflow.com/a/9229821
export const uniq_fast = (a: string[]): string[] => {
  const seen: Record<string, number> = {};
  const out = [];
  const len = a.length;
  let j = 0;
  for (let i = 0; i < len; i++) {
    const item = a[i];
    if (seen[item] !== 1) {
      seen[item] = 1;
      out[j++] = item;
    }
  }
  return out;
};

// The goal for this recursive function is to check to ensure
// that the keys are held down in the correct order of the shortcut.
// I.E if the shortcut array is ["Shift", "E", "A"], this function will ensure
// that "E" is held down before "A", and "Shift" is held down before "E".
export const checkHeldKeysRecursive = (
  shortcutKey: string,
  // Tracks the call interation for the recursive function,
  // based on the previous index;
  shortcutKeyRecursionIndex = 0,
  shortcutArray: string[],
  heldKeysArray: string[],
): boolean => {
  const shortcutIndexOfKey = shortcutArray.indexOf(shortcutKey);
  const keyPartOfShortCut = shortcutArray.indexOf(shortcutKey) >= 0;

  // Early exit if they key isn't even in the shortcut combination.
  if (!keyPartOfShortCut) {
    return false;
  }

  // While holding down one of the keys, if another is to be let go, the shortcut
  // should be void. Shortcut keys must be held down in a specifc order.
  // This function is always called before a key is added to held keys on keydown,
  // this will ensure that heldKeys only contains the prefixing keys
  const comparisonIndex = Math.max(heldKeysArray.length - 1, 0);
  if (heldKeysArray.length && heldKeysArray[comparisonIndex] !== shortcutArray[comparisonIndex]) {
    return false;
  }

  // Early exit for the first held down key in the shortcut,
  // except if this is a recursive call
  if (shortcutIndexOfKey === 0) {
    // If this isn't the first interation of this recursive function, and we're
    // recursively calling this function, we should always be checking the
    // currently held down keys instead of returning true
    if (shortcutKeyRecursionIndex > 0) {
      return heldKeysArray.indexOf(shortcutKey) >= 0;
    }
    return true;
  }

  const previousShortcutKeyIndex = shortcutIndexOfKey - 1;
  const previousShortcutKey = shortcutArray[previousShortcutKeyIndex];
  const previousShortcutKeyHeld = heldKeysArray[previousShortcutKeyIndex] === previousShortcutKey;

  // Early exit if the key just before the currently checked shortcut key
  // isn't being held down.
  if (!previousShortcutKeyHeld) {
    return false;
  }

  // Recursively call this function with the previous key as the new shortcut key
  // but the index of the current shortcut key.
  return checkHeldKeysRecursive(previousShortcutKey, shortcutIndexOfKey, shortcutArray, heldKeysArray);
};
