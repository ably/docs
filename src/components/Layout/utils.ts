import cn from '@ably/ui/core/utils/cn';
import { AccordionProps } from '@ably/ui/core/Accordion';
import { ProductData, ProductKey } from 'src/data/types';
import { NavProductContent, NavProductPage, NavProductPages } from 'src/data/nav/types';

export type PageTreeNode = { index: number; page: NavProductPage };

type ActivePage = { tree: PageTreeNode[]; page: NavProductPage };

/**
 * Determines the active page based on the provided target link.
 *
 * This function iterates through the product data to find a page that matches the target link.
 * It checks both direct links and nested pages within groups. If a match is found, it returns
 * the index tree of the page in the navigation data and the page details.
 *
 * @param data - The product data containing navigation information.
 * @param targetLink - The link to be matched against the product data.
 * @returns The active page details including its tree and page information, or null if no match is found.
 */
export const determineActivePage = (data: ProductData, targetLink: string): ActivePage | null => {
  const strippedTargetLink = formatNavLink(targetLink);

  /**
   * Recursively determines the presence of a page in a list of pages.
   *
   * This function iterates through the list of pages and checks if the target link is present.
   * If a match is found, it returns the index of the link in its pages array.
   * If not, it attempts to search for the link in nested pages.
   */
  const determinePagePresence = (pages: NavProductPages[], path: PageTreeNode[]): PageTreeNode[] | null => {
    for (let i = 0; i < pages.length; i++) {
      const page = pages[i];

      // If the page is a link and the link matches the target link, return the path
      if ('link' in page && formatNavLink(page.link) === strippedTargetLink) {
        return [...path, { index: i, page }];
      }

      // If the page is a group of pages, recursively check each page
      if ('pages' in page && page.pages) {
        const result = determinePagePresence(page.pages, [...path, { index: i, page: { ...page, link: '#' } }]);
        if (result) {
          return result;
        }
      }
    }
    return null;
  };

  for (const key of Object.keys(data) as ProductKey[]) {
    /* 
      If the page matches a product index link, return details for the product
    */
    if (data[key].nav.link === strippedTargetLink) {
      const { name, link } = data[key].nav;
      return { tree: [{ index: Object.keys(data).indexOf(key), page: { name, link } }], page: { name, link } };
    }

    /* 
      If not, attempt to calculate page tree hierarchies as both content and API pages
      and return the first valid tree result along with an object containing details
      on the matched page
    */
    if (data[key].nav.content) {
      const contentResult = determinePagePresence(data[key].nav.content, []);
      const apiResult = determinePagePresence(data[key].nav.api, []);
      if (contentResult || apiResult) {
        const tree = [
          {
            index: Object.keys(data).indexOf(key),
            page: { ...data[key].nav, link: data[key].nav.link ?? '#' },
          },
          ...((contentResult || apiResult) ?? []),
        ];
        const page = tree.slice(1).reduce<NavProductPages[]>(
          (acc, curr) => {
            if (acc[curr.index] && 'pages' in acc[curr.index]) {
              return (acc[curr.index] as NavProductContent).pages;
            }

            return [acc[curr.index]];
          },
          data[key].nav[apiResult ? 'api' : 'content'],
        );

        return { tree, page: page?.[0] as NavProductPage };
      }
    }
  }

  return null;
};

export const formatNavLink = (link: string) => {
  if (!link) {
    return '';
  }

  if (link.startsWith('/docs')) {
    link = link.replace('/docs', '');
  }

  return link.replace(/\/$/, '');
};

export const commonAccordionOptions = (
  currentPage: NavProductContent | null,
  openIndex: number | undefined,
  topLevel: boolean,
  inHeader: boolean,
): Omit<AccordionProps, 'data'> => ({
  icons: { open: { name: 'icon-gui-chevron-up' }, closed: { name: 'icon-gui-chevron-down' } },
  options: {
    autoClose: topLevel,
    headerCSS: cn(
      'text-neutral-1000 dark:text-neutral-300 md:text-neutral-900 dark:md:text-neutral-400 hover:text-neutral-1100 active:text-neutral-1000 !py-0 pl-0 !mb-0 transition-colors [&_svg]:!w-24 [&_svg]:!h-24 md:[&_svg]:!w-20 md:[&_svg]:!h-20',
      {
        'my-12': topLevel && inHeader,
        'h-40 ui-text-menu1 font-bold md:ui-text-menu4 px-16': topLevel,
        'h-[1rem] ui-text-menu2 font-semibold md:ui-text-menu4': !topLevel,
      },
    ),
    selectedHeaderCSS: 'text-neutral-1300 mb-8',
    contentCSS: cn('[&>div]:pb-0'),
    rowIconSize: '20px',
    defaultOpenIndexes: !inHeader && openIndex !== undefined ? [openIndex] : [],
    hideBorders: true,
    fullyOpen: !topLevel && currentPage?.expand,
  },
});

export const sidebarAlignmentClasses = 'absolute md:sticky w-240 md:pb-128 top-[88px] h-[calc(100vh-88px)]';

export const composeNavLinkId = (link: string) => `nav-link-${formatNavLink(link).replaceAll('/', '-')}`;

export const hierarchicalKey = (id: string, depth: number, tree?: PageTreeNode[]) =>
  [...(tree ? tree.slice(0, depth).map((node) => node.index) : []), id].join('-');

export const pathWithBase = (path: string) => {
  const strippedPath = path.replace(/^\/+/, '');
  return `/${process.env.NODE_ENV === 'development' ? '' : 'docs/'}${strippedPath}`;
};
