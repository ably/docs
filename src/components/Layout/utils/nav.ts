import cn from '@ably/ui/core/utils/cn';
import { AccordionProps } from '@ably/ui/core/Accordion';
import { HEADER_HEIGHT, componentMaxHeight } from '@ably/ui/core/utils/heights';
import { ProductData, ProductKey } from 'src/data/types';
import { NavProductContent, NavProductPage, NavProductPages } from 'src/data/nav/types';
import { LanguageKey } from 'src/data/languages/types';
import { DEFAULT_LANGUAGE } from 'src/contexts/layout-context';

export type PageTreeNode = { index: number; page: NavProductPage };

export type ActivePage = {
  tree: PageTreeNode[];
  page: NavProductPage;
  languages: LanguageKey[];
  language: LanguageKey;
  product: ProductKey | null;
};

/**
 * Determines the active page based on the provided target link.
 *
 * This function iterates through the product data to find a page that matches the target link.
 * It checks both direct links and nested pages within groups. If a match is found, it returns
 * the index tree of the page in the navigation data and the page details.
 *
 * @param data - The product data containing navigation information.
 * @param targetLink - The link to be matched against the product data.
 * @returns The active page details including its tree and page information, or null if no match is found.
 */
export const determineActivePage = (data: ProductData, targetLink: string): ActivePage | null => {
  const strippedTargetLink = formatNavLink(targetLink);

  /**
   * Recursively determines the presence of a page in a list of pages.
   *
   * This function iterates through the list of pages and checks if the target link is present.
   * If a match is found, it returns the index of the link in its pages array.
   * If not, it attempts to search for the link in nested pages.
   */
  const determinePagePresence = (pages: NavProductPages[], path: PageTreeNode[]): PageTreeNode[] | null => {
    for (let i = 0; i < pages.length; i++) {
      const page = pages[i];

      // If the page is a link and the link matches the target link, return the path
      if ('link' in page && formatNavLink(page.link) === strippedTargetLink) {
        return [...path, { index: i, page }];
      }

      // If the page is a group of pages, recursively check each page
      if ('pages' in page && page.pages) {
        const result = determinePagePresence(page.pages, [...path, { index: i, page: { ...page, link: '#' } }]);
        if (result) {
          return result;
        }
      }
    }
    return null;
  };

  for (const key of Object.keys(data) as ProductKey[]) {
    /* 
      If the page matches a product index link, return details for the product
    */
    if (!data[key].nav) {
      continue;
    }

    if (data[key].nav.link === strippedTargetLink) {
      const { name, link } = data[key].nav;
      return {
        tree: [{ index: Object.keys(data).indexOf(key), page: { name, link } }],
        page: { name, link },
        languages: [],
        language: DEFAULT_LANGUAGE,
        product: key,
      };
    }

    /* 
      If not, attempt to calculate page tree hierarchies as both content and API pages
      and return the first valid tree result along with an object containing details
      on the matched page
    */
    if (data[key].nav.content) {
      const contentResult = determinePagePresence(data[key].nav.content, []);
      const apiResult = determinePagePresence(data[key].nav.api, []);
      if (contentResult || apiResult) {
        const tree = [
          {
            index: Object.keys(data).indexOf(key),
            page: { ...data[key].nav, link: data[key].nav.link ?? '#' },
          },
          ...((contentResult || apiResult) ?? []),
        ];
        const page = tree.slice(1).reduce<NavProductPages[]>(
          (acc, curr) => {
            if (acc[curr.index] && 'pages' in acc[curr.index]) {
              return (acc[curr.index] as NavProductContent).pages;
            }

            return [acc[curr.index]];
          },
          data[key].nav[apiResult ? 'api' : 'content'],
        );

        return { tree, page: page?.[0] as NavProductPage, languages: [], language: DEFAULT_LANGUAGE, product: key };
      }
    }
  }

  return null;
};

export const formatNavLink = (link: string) => {
  if (!link) {
    return '';
  }

  if (link.startsWith('/docs')) {
    link = link.replace('/docs', '');
  }

  // Strip out query params
  link = link.split('?')[0];

  return link.replace(/\/$/, '');
};

export const commonAccordionOptions = (
  currentPage: NavProductContent | null,
  openIndex: number | undefined,
  topLevel: boolean,
  inHeader: boolean,
): Omit<AccordionProps, 'data'> => ({
  icons: { open: { name: 'icon-gui-chevron-up-micro' }, closed: { name: 'icon-gui-chevron-down-micro' } },
  options: {
    autoClose: topLevel,
    headerCSS: cn(
      'text-neutral-1000 dark:text-neutral-300 md:text-neutral-900 dark:md:text-neutral-400 hover:text-neutral-1100 active:text-neutral-1000 !py-0 pl-0 !mb-0 transition-colors [&_svg]:!w-24 [&_svg]:!h-24 md:[&_svg]:!w-20 md:[&_svg]:!h-20',
      {
        'my-12': topLevel && inHeader,
        'h-40 ui-text-label1 !font-bold md:ui-text-label4 px-16': topLevel,
        'min-h-[1.625em] md:min-h-[1.375em] ui-text-label2 !font-semibold md:ui-text-label4': !topLevel,
      },
    ),
    selectedHeaderCSS: '!text-neutral-1300 mb-8',
    contentCSS: '[&>div]:pb-0',
    rowIconSize: '20px',
    defaultOpenIndexes: !inHeader && openIndex !== undefined ? [openIndex] : [],
    hideBorders: true,
    fullyOpen: !topLevel && currentPage?.expand,
  },
});

export const sidebarAlignmentClasses = 'absolute md:sticky w-[240px] md:pb-128 pt-24';

export const sidebarAlignmentStyles: React.CSSProperties = {
  top: HEADER_HEIGHT,
  height: componentMaxHeight(HEADER_HEIGHT),
};

export const composeNavLinkId = (link: string) => `nav-link-${formatNavLink(link).replaceAll('/', '-')}`;

export const hierarchicalKey = (id: string, depth: number, tree?: PageTreeNode[]) =>
  [...(tree ? tree.slice(0, depth).map((node) => node.index) : []), id].join('-');
