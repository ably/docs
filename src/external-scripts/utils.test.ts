import { scriptLoader } from './utils';

describe('Utils', () => {
  describe('scriptLoader', () => {
    beforeEach(() => {
      document.body.innerHTML = `<script></script>`;
    });

    it('attaches a script tag', () => {
      scriptLoader(document, 'http://ga.com', { 'data-testid': 'script' });
      expect(document.body).toMatchSnapshot();
    });

    it('attaches a script tag with custom attributes', () => {
      scriptLoader(document, 'http://ga.com', { title: 'ga' });
      expect(document.body).toMatchSnapshot();
    });

    it('can replace defaults with custom attributes', () => {
      scriptLoader(document, 'http://ga.com', { async: false, defer: false, type: 'module' });
      expect(document.body).toMatchSnapshot();

      const script = document.querySelector('script[src="http://ga.com"]');
      expect(script).not.toBeNull();
      expect(script.async).toBeFalsy();
      expect(script.defer).toBeFalsy();
      expect(script.type).toBe('module');
    });

    it('can add data attributes', () => {
      scriptLoader(document, 'http://example.com', { data: { test: '1' } });

      const script = document.querySelector('script[src="http://example.com"]');
      expect(script).not.toBeNull();
      expect(script.dataset.test).toBe('1');
    });
  });
});
