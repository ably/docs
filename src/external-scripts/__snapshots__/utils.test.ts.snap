// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Utils scriptLoader attaches a script tag 1`] = `
<body>
  <script
    defer=""
    src="http://ga.com"
    type="text/javascript"
  />
  <script />
</body>
`;

exports[`Utils scriptLoader attaches a script tag with custom attributes 1`] = `
<body>
  <script
    defer=""
    src="http://ga.com"
    title="ga"
    type="text/javascript"
  />
  <script />
</body>
`;

exports[`Utils scriptLoader can replace defaults with custom attributes 1`] = `
<body>
  <script
    src="http://ga.com"
    type="module"
  />
  <script />
</body>
`;
